torch/jit/__init__.py

```diff
@@ -1731,6 +1731,31 @@ def __getstate__(self):
                 "Mixed serialization of script and non-script modules is not supported. " +
                 "For purely script modules use my_script_module.save(<filename>) instead.")
 
+        # Python magic methods do lookups on an object's class type, instead of looking up
+        # the desugared attribute. In order to continue to expose the magic methods
+        # of builtin-containers (ModuleList, Sequential, ModuleDict) to python we
+        # define magic methods here as a shim to the correct attribute.
+        def forward_magic_method(self, method_name, *args, **kwargs):
+            self_method = getattr(self, method_name)
+            if self_method == getattr(RecursiveScriptModule, method_name):
+                raise NotImplementedError()
+            return self_method(*args, **kwargs)
+
+        def __iter__(self):
+            return self.forward_magic_method("__iter__")
+
+        def __getitem__(self, idx):
+            return self.forward_magic_method("__getitem__", idx)
+
+        def __len__(self):
+            return self.forward_magic_method("__len__")
+
+        def __dir__(self):
```

Comment:
`nn.Module` implements `__dir__`, so I think it breaks for regular modules.

Reply:
So in the case of `__dir__` we should always `@_copy_to_script` so to speak ? or what should be done here

Reply:
Since `RecursiveScriptModule` inherits from `nn.Module`, `__dir__` is always defined on `RecursiveScriptModule`s. This makes `forward_magic_method` problematic, because it takes over the default behavior. I think the "generate a new type" strategy would work to fix this, since it would just further override `__dir__`.